proxy/proxy_test.go

```diff
@@ -1451,6 +1546,145 @@ var _ = Describe("Proxy", func() {
 
 					}, "20s").Should(Succeed()) // we don't know which endpoint will be chosen first, so we have to try until sequence 1,2,3 has been hit
 				})
+				It("shows no backend_time or other attempt details if all endpoints are broken", func() {
+
+					// Register some broken endpoints to cause retries
+					test_util.RegisterAddr(r, "fully-broken-app", "10.255.255.1:1234", test_util.RegisterConfig{InstanceIndex: "1"})
+					test_util.RegisterAddr(r, "fully-broken-app", "10.255.255.1:1235", test_util.RegisterConfig{InstanceIndex: "2"})
+					test_util.RegisterAddr(r, "fully-broken-app", "10.255.255.1:1236", test_util.RegisterConfig{InstanceIndex: "3"})
+
+					conn := dialProxy(proxyServer)
+					req := test_util.NewRequest("GET", "fully-broken-app", "/", nil)
+
+					started := time.Now()
+					conn.WriteRequest(req)
+
+					resp, err := http.ReadResponse(conn.Reader, &http.Request{})
+					Expect(err).NotTo(HaveOccurred())
+
+					Expect(resp.StatusCode).To(Equal(http.StatusBadGateway))
+					Expect(time.Since(started)).To(BeNumerically("<", 5*time.Second))
+
+					Eventually(func() (int64, error) {
+						fi, err := f.Stat()
+						if err != nil {
+							return 0, err
+						}
+						return fi.Size(), nil
+					}, "10s").ShouldNot(BeZero())
+
+					logBytes, err := os.ReadFile(f.Name())
+					Expect(err).NotTo(HaveOccurred())
+					logStr := string(logBytes)
+					Expect(strings.HasPrefix(logStr, "fully-broken-app - [")).To(BeTrue())
+					Expect(logStr).To(ContainSubstring("response_time:"))
+					Expect(logStr).To(ContainSubstring("gorouter_time:"))
+					Expect(logStr).To(MatchRegexp("failed_attempts:3"))
+					Expect(logStr).To(MatchRegexp("failed_attempts_time:3"))
+					Expect(logStr).To(MatchRegexp("response_time:3"))
+					Expect(logStr).To(MatchRegexp("gorouter_time:0"))
+					Expect(logStr).To(MatchRegexp(`backend_time:"-"`))
+					Expect(logStr).To(MatchRegexp(`dns_time:"-"`))
+					Expect(logStr).To(MatchRegexp(`dial_time:"-"`))
+					Expect(logStr).To(MatchRegexp(`tls_time:"-"`))
+				})
+				Context("in websocket requests", func() {
+					BeforeEach(func() {
+						conf.WebsocketDialTimeout = 1 * time.Second
+					})
+					It("shows slowness in failed_attempts_time and backend_time, not gorouter_time", func() {
+
+						// Register some broken endpoints to cause retries
+						test_util.RegisterAddr(r, "partially-broken-ws-app", "10.255.255.1:1234", test_util.RegisterConfig{InstanceIndex: "1"})
+						test_util.RegisterAddr(r, "partially-broken-ws-app", "10.255.255.1:1235", test_util.RegisterConfig{InstanceIndex: "2"})
+
+						ln := test_util.RegisterWSHandler(r, "partially-broken-ws-app", func(conn *websocket.Conn) {
+							msgBuf := make([]byte, 100)
+							n, err := conn.Read(msgBuf)
+							Expect(err).NotTo(HaveOccurred())
+							Expect(string(msgBuf[:n])).To(Equal("HELLO WEBSOCKET"))
+
+							time.Sleep(2 * time.Second)
```

Comment:
Similar to above, I'd like to lessen this sleep. I see why you did it because you are testing these different times. Any chance you can lower it to some fraction of a second instead of 2 seconds?